node-js-basics-101/node_modules/ws/lib/PerMessageDeflate.js

```diff
@@ -0,0 +1,384 @@
+    return paramsList.map((params) => {
+      Object.keys(params).forEach((key) => {
+        var value = params[key];
+        if (value.length > 1) {
+          throw new Error(`Multiple extension parameters for ${key}`);
+        }
+
+        value = value[0];
+
+        switch (key) {
+          case 'server_no_context_takeover':
+          case 'client_no_context_takeover':
+            if (value !== true) {
+              throw new Error(`invalid extension parameter value for ${key} (${value})`);
+            }
+            params[key] = true;
+            break;
+          case 'server_max_window_bits':
+          case 'client_max_window_bits':
+            if (typeof value === 'string') {
+              value = parseInt(value, 10);
+              if (!~AVAILABLE_WINDOW_BITS.indexOf(value)) {
+                throw new Error(`invalid extension parameter value for ${key} (${value})`);
+              }
+            }
+            if (!this._isServer && value === true) {
+              throw new Error(`Missing extension parameter value for ${key}`);
+            }
+            params[key] = value;
+            break;
+          default:
+            throw new Error(`Not defined extension parameter (${key})`);
+        }
+      });
+      return params;
+    });
+  }
+
+  /**
+   * Decompress data.
+   *
+   * @param {Buffer} data Compressed data
+   * @param {Boolean} fin Specifies whether or not this is the last fragment
+   * @param {Function} callback Callback
+   * @public
+   */
+  decompress (data, fin, callback) {
+    const endpoint = this._isServer ? 'client' : 'server';
+
+    if (!this._inflate) {
+      const maxWindowBits = this.params[`${endpoint}_max_window_bits`];
+      this._inflate = zlib.createInflateRaw({
+        windowBits: typeof maxWindowBits === 'number' ? maxWindowBits : DEFAULT_WINDOW_BITS
+      });
+    }
+    this._inflate.writeInProgress = true;
+
+    var totalLength = 0;
+    const buffers = [];
+    var err;
+
+    const onData = (data) => {
+      totalLength += data.length;
+      if (this._maxPayload < 1 || totalLength <= this._maxPayload) {
+        return buffers.push(data);
+      }
+
+      err = new Error('max payload size exceeded');
+      err.closeCode = 1009;
+      this._inflate.reset();
+    };
+
+    const onError = (err) => {
+      cleanup();
+      callback(err);
+    };
+
+    const cleanup = () => {
+      if (!this._inflate) return;
+
+      this._inflate.removeListener('error', onError);
+      this._inflate.removeListener('data', onData);
+      this._inflate.writeInProgress = false;
+
+      if (
+        (fin && this.params[`${endpoint}_no_context_takeover`]) ||
+        this._inflate.pendingClose
+      ) {
+        this._inflate.close();
+        this._inflate = null;
+      }
+    };
+
+    this._inflate.on('error', onError).on('data', onData);
+    this._inflate.write(data);
+    if (fin) this._inflate.write(TRAILER);
+
+    this._inflate.flush(() => {
```

Comment:
__[jshint]__
* 'arrow function syntax (=>)' is only available in ES6 (use 'esversion: 6').

<!-- Comment by [SideCI](https://sideci.com) -->